File: java/src/main/java/com/myriad/wdl/model/resolvers/WdlImportResolverBase.java
package com.myriad.wdl.model.resolvers;

import com.myriad.wdl.model.errors.WdlImportException;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Base import resolver that dispatches URI-like import locations to protocol-specific handlers.
 *
 * <p>Built-in dispatch covers WDL 1.x protocols ({@code http://}, {@code https://}, and
 * deprecated {@code file://}) plus protocol-less paths resolved relative to the current document
 * location. Concrete subclasses provide transport-specific HTTP(S) behavior.
 */
public abstract class WdlImportResolverBase {

  /**
   * Resolves and loads import content.
   *
   * @param currentDocumentLocation location of the document that contains the import statement;
   *     may be null if unknown
   * @param importLocation import URI/path string from the WDL source
   */
  public String resolveImport(URI currentDocumentLocation, String importLocation)
      throws WdlImportException {
    URI resolved = resolveImportLocation(currentDocumentLocation, importLocation);
    return dispatchImport(currentDocumentLocation, resolved, importLocation);
  }

  /** Resolves an import location string into a URI identifier used by loader import maps. */
  public URI resolveImportLocation(URI currentDocumentLocation, String importLocation)
      throws WdlImportException {
    return resolveImportUri(currentDocumentLocation, importLocation);
  }

  /**
   * Dispatches a resolved import URI to a concrete loader.
   *
   * <p>Subclasses can override to support non-standard protocols. A typical pattern is to handle
   * custom schemes first, then call {@code super.dispatchImport(...)} for spec-compliant behavior.
   */
  protected String dispatchImport(
      URI currentDocumentLocation, URI importUri, String originalImportLocation)
      throws WdlImportException {
    String scheme = importUri.getScheme();
    if (scheme == null) {
      return loadBarePath(currentDocumentLocation, importUri.getPath(), originalImportLocation);
    }

    switch (scheme.toLowerCase()) {
      case "http":
        return loadHttp(currentDocumentLocation, importUri, originalImportLocation);
      case "https":
        return loadHttps(currentDocumentLocation, importUri, originalImportLocation);
      case "file":
        return loadFile(currentDocumentLocation, importUri, originalImportLocation);
      default:
        throw new WdlImportException(
            "Unsupported import URI protocol: " + importUri.getScheme(), originalImportLocation);
    }
  }

  protected abstract String loadHttp(
      URI currentDocumentLocation, URI importUri, String originalImportLocation)
      throws WdlImportException;

  protected abstract String loadHttps(
      URI currentDocumentLocation, URI importUri, String originalImportLocation)
      throws WdlImportException;

  protected String loadFile(
      URI currentDocumentLocation, URI importUri, String originalImportLocation)
      throws WdlImportException {
    try {
      return Files.readString(Path.of(importUri), StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw ioFailure("Unable to read file import", originalImportLocation, e);
    }
  }

  protected String loadBarePath(
      URI currentDocumentLocation, String resolvedPath, String originalImportLocation)
      throws WdlImportException {
    if (resolvedPath == null || resolvedPath.isBlank()) {
      throw new WdlImportException("Invalid filesystem import path", originalImportLocation);
    }
    try {
      return Files.readString(Path.of(resolvedPath), StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw ioFailure("Unable to read filesystem import", originalImportLocation, e);
    }
  }

  protected URI resolveImportUri(URI currentDocumentLocation, String importLocation)
      throws WdlImportException {
    try {
      URI importUri = new URI(importLocation);
      if (importUri.getScheme() != null) {
        return importUri;
      }

      if (currentDocumentLocation == null) {
        if (importLocation.startsWith("/")) {
          return new URI("file", null, importLocation, null);
        }
        return importUri;
      }

      if (importLocation.startsWith("/")) {
        return new URI(
            currentDocumentLocation.getScheme(),
            currentDocumentLocation.getAuthority(),
            importLocation,
            null,
            null);
      }

      URI base = currentDocumentLocation;
      if (base.getPath() != null && !base.getPath().endsWith("/")) {
        base = base.resolve(".");
      }
      return base.resolve(importUri);
    } catch (URISyntaxException e) {
      throw new WdlImportException("Invalid import URI", importLocation, e);
    }
  }

  protected static WdlImportException ioFailure(
      String message, String importLocation, IOException cause) {
    return new WdlImportException(message, importLocation, cause);
  }
}

File: typescript/src/model/resolvers/wdl-import-resolver-base.ts
import { WdlImportException } from '../errors/wdl-import-exception.js';

/** Transport-agnostic base resolver for WDL imports. */
export abstract class WdlImportResolverBase {
  /** Resolves an import reference and loads its source content. */
  public resolveImport(
    currentDocumentLocation: string | undefined,
    importLocation: string,
  ): string {
    const resolved = this.resolveImportLocation(currentDocumentLocation, importLocation);
    return this.dispatchImport(currentDocumentLocation, resolved, importLocation);
  }

  /** Resolves an import reference into a stable location identifier. */
  public resolveImportLocation(
    currentDocumentLocation: string | undefined,
    importLocation: string,
  ): string {
    return this.resolveImportUri(currentDocumentLocation, importLocation);
  }

  /** Dispatches loading behavior based on URI protocol. */
  protected dispatchImport(
    currentDocumentLocation: string | undefined,
    importUri: string,
    originalImportLocation: string,
  ): string {
    const parsed = this.tryParseUrl(importUri);
    const scheme = parsed?.protocol?.replace(':', '').toLowerCase() ?? '';

    if (!scheme) {
      return this.loadBarePath(currentDocumentLocation, importUri, originalImportLocation);
    }
    if (scheme === 'http') {
      return this.loadHttp(currentDocumentLocation, importUri, originalImportLocation);
    }
    if (scheme === 'https') {
      return this.loadHttps(currentDocumentLocation, importUri, originalImportLocation);
    }
    if (scheme === 'file') {
      return this.loadFile(currentDocumentLocation, importUri, originalImportLocation);
    }
    throw new WdlImportException(
      `Unsupported import URI protocol: ${scheme}`,
      originalImportLocation,
    );
  }

  protected abstract loadHttp(
    currentDocumentLocation: string | undefined,
    importUri: string,
    originalImportLocation: string,
  ): string;

  protected abstract loadHttps(
    currentDocumentLocation: string | undefined,
    importUri: string,
    originalImportLocation: string,
  ): string;

  protected abstract loadFile(
    currentDocumentLocation: string | undefined,
    importUri: string,
    originalImportLocation: string,
  ): string;

  protected abstract loadBarePath(
    currentDocumentLocation: string | undefined,
    resolvedPath: string,
    originalImportLocation: string,
  ): string;

  protected resolveImportUri(
    currentDocumentLocation: string | undefined,
    importLocation: string,
  ): string {
    const parsedImport = this.tryParseUrl(importLocation);
    if (parsedImport?.protocol) {
      return importLocation;
    }

    if (!currentDocumentLocation) {
      if (importLocation.startsWith('/')) {
        return new URL(`file://${importLocation}`).toString();
      }
      return importLocation;
    }

    const base = this.tryParseUrl(currentDocumentLocation);
    const baseScheme = base?.protocol?.replace(':', '').toLowerCase();

    if (baseScheme === 'http' || baseScheme === 'https') {
      return new URL(importLocation, currentDocumentLocation).toString();
    }

    if (baseScheme === 'file') {
      return new URL(importLocation, currentDocumentLocation).toString();
    }

    const baseDir = this.dirname(currentDocumentLocation);
    return this.normalizePathJoin(baseDir, importLocation);
  }

  protected ioFailure(message: string, importLocation: string, cause: unknown): WdlImportException {
    return new WdlImportException(message, importLocation, cause);
  }

  private tryParseUrl(value: string): URL | undefined {
    try {
      return new URL(value);
    } catch {
      return undefined;
    }
  }

  private dirname(path: string): string {
    const normalized = path.replace(/\\/g, '/');
    const idx = normalized.lastIndexOf('/');
    if (idx < 0) return '';
    return normalized.slice(0, idx);
  }

  private normalizePathJoin(baseDir: string, child: string): string {
    if (!baseDir) return child;
    if (child.startsWith('/')) return child;
    return `${baseDir}/${child}`.replace(/\/+/g, '/');
  }
}
